### PR TITLE
WPP-139 - Profile Picture Upload Backend

### PR DESCRIPTION
**Pull request by PistolRcks**
## For Pull Requesters/Reviewers:
All of the following should be true before making a pull request and should be inspected by reviewers:

### User Story
- [X] Acceptance criteria are adhered to and completely finished
- [X] All subtasks on Jira are completed

### Unit Testing
- [X] Unit tests are finished and passing
- [X] Code coverage is >80% (so much branch testing....)

### Documentation/Style
- [X] JSDocs are written for methods and components
- [X] Unclear code has been documented
- [X] Style standards are adhered to

## For Reviewers:
The following should be additionally checked by reviewers:

- [x] Code runs locally
- [x] Visually inspect tests; they should not simply pass, but rather they should be doing something and testing acceptance criteria
- [x] See if "jank" code could be rewritten

## Description
Adds a POST route, `/api/user/picture`, which updates the user's profile picture and runs some validation on it. Also, the previous POST route, `/api/user`, has been changed to `/api/user/password` for clarity.

And there's a *lot* of branch testing.

### CODE REVIEW

File: server/api/userInfo.js [hunk 1]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
+  
+  // Type validation
+  const { profile_picture, profile_picture_top, profile_picture_left } = req.body
+  
+  if (isNaN(profile_picture_top) || isNaN(profile_picture_left)) {
+    res.status(400).send("Error: 'profile_picture_top' or 'profile_picture_left' is not a number!")
+  }
+  
+  // Space validation
+  // TODO: Maybe we should figure out how we could make sure the cropping is not greater than the size of the image?
```
Comment (Blackkirby72) [inline server/api/userInfo.js:234]: more easily done if the endpoint were to accept an image file as opposed to an already converted base64 encoded string

File: server/api/userInfo.js [hunk 2]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
+ *  ```js
+ *  {
+ *    "profile_picture": "A string-encoded binary blob of the image being uploaded",
+ *    "profile_picture_top": the integer cropping point from the top of the image,
+ *    "profile_picture_left": the integer cropping point from the left of the image,
+ *  }
+ *  ```
+ * @param {Express.Response} res - The Response object generated by the route.
+ * @param {Function} next - The next() function used in the callback.
+ * @returns Nothing.
```
Comment (Blackkirby72) [inline server/api/userInfo.js]: I'd just remove the `@returns` if it returns nothing

**Review by prewe5294 — APPROVED** [review]
Looks good, as far as I know. Code runs locally and tests pass

File: server/api/userInfo.js [hunk 3]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
 
+/**
+ * The callback function for the /api/user/picture POST route. Updates a User's profile picture.
+ *
+ * This endpoint requires authorization.
+ * @param {Express.Request} req - The Request object generated by the route
+ *  The request body should be formatted as JSON with the following:
+ *  ```js
+ *  {
+ *    "profile_picture": "A string-encoded binary blob of the image being uploaded",
```
Comment (Blackkirby72) [inline server/api/userInfo.js:189]: update to `base64 encoded string`

File: server/api/userInfo.js [hunk 4]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
+      .send(
+        "Error: 'profile_picture', 'profile_picture_top', or 'profile_picture_left' not found in the request body!"
+      );
+    return;
+  }
+  
+  // Type validation
+  const { profile_picture, profile_picture_top, profile_picture_left } = req.body
+  
+  if (isNaN(profile_picture_top) || isNaN(profile_picture_left)) {
```
Comment (Blackkirby72) [inline server/api/userInfo.js:218]: can we use some regex to validate that `profile_picture` only contains `a-zA-Z0-9+/` (the base64 characters)?

File: server/api/userInfo.js [hunk 5]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
+    res.status(400).send("Error: 'profile_picture_top' or 'profile_picture_left' is not a number!")
```
Comment (Blackkirby72) [inline server/api/userInfo.js]: need to `return`, hitting this spot currently crashes the server

File: server/api/userInfo.js [hunk 6]
```diff
@@ -179,7 +179,68 @@ function postUserInfo(req, res, next) {
...
+  }
+  
+  // Space validation
+  // TODO: Maybe we should figure out how we could make sure the cropping is not greater than the size of the image?
+  if (profile_picture_top < 0 || profile_picture_left < 0) {
+    res.status(400).send("Error: 'profile_picture_top' or 'profile_picture_left' is negative!")
```
Comment (Blackkirby72) [inline server/api/userInfo.js]: also crashes the server, needs `return`

**Review by Blackkirby72 — CHANGES_REQUESTED** [review]

File: server/api/userInfo.js (hunk 1 above)
Comment (PistolRcks) [reply server/api/userInfo.js:234]: ehh, it's not required. would be too much work for me to research, so I'm not gonna 🤷

Comment (Blackkirby72) [reply server/api/userInfo.js:234]: good enough for now

**Review by Blackkirby72 — APPROVED** [review]

**Pull request merged**